File: python-ds-practice/34_same_frequency/same_frequency.py
def freq_counter(num):
    """Returns a frequency counter mapping of a number"""

    counts = {}

    for digit in num:
        counts[digit] = counts.get(digit, 0) + 1

    return counts


def same_frequency(num1, num2):
    """Do these nums have same frequencies of digits?

        >>> same_frequency(551122, 221515)
        True

        >>> same_frequency(321142, 3212215)
        False

        >>> same_frequency(1212, 2211)
        True
    """
    return freq_counter(str(num1)) == freq_counter(str(num2))

File: prob-freq-count-pointers/same-frequency.js
// input: two positive integers

//output: boolean => true if both numbers have the same frequency of digits, otherwise false

//constraint: O(n + m) so sorting not allowed

function sameFrequency(num1, num2) {
  let num1Arr = num1.toString().split("");
  let num2Arr = num2.toString().split("");

  //cant have same freq of nums if not same length!
  if (num1Arr.length !== num2Arr.length) return false;

  let num1Freq = {};
  let num2Freq = {};

  for (let num of num1Arr) {
    num1Freq[num] = num1Freq[num] + 1 || 1;
  }
  for (let num of num2Arr) {
    num2Freq[num] = num2Freq[num] + 1 || 1;
  }

  console.log(num1Freq);
  console.log(num2Freq);

  for (let key in num1Freq) {
    if (num1Freq[key] !== num2Freq[key]) return false;
  }

  return true;
}

module.exports = sameFrequency;
